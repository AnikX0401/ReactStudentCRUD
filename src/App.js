import { useState, useEffect } from "react";
import "./index.css";
import {
  getStudents,
  addStudents,
  deleteStudent,
  editStudent,
} from "./components/service";
import Students from "./components/Students";
import AddStudent from "./components/AddStudent";
import EditStudent from "./components/EditStudent";

function App() {
  const [students, setStudents] = useState([]);
  const [showAddStudentForm, setshowAddStudentForm] = useState(false);
  const [refresh, setRefresh] = useState(false);
  const [showEditStudentForm, setshowEditStudentForm] = useState(false);
  const [studentToEdit, setStudentToEdit] = useState();
  const [error, setError] = useState(false);

  useEffect(() => {
    console.log("customStu::::::", students);
    getStudents().then((res) => {
      console.log("res::", res);

      if (res?.error === "failed") {
        setError(true);
      } else {
        setError(false);
        setStudents(res);
      }
    });
  }, [refresh]);

  console.log("mystudents:::", students);
  // const changeStudentName = () => {
  //   const newStudentList = students.map((student, index) => {
  //     if (index < 0) {
  //       student.name = "New name";
  //       student.rollNumber = "New rollNumber";
  //     } else {
  //       if (index > 0) {
  //         student.name = "New Student";
  //         student.rollNumber = "rollNumber";
  //       }
  //     }
  //     return student;
  //   });
  //   setStudents(newStudentList);
  // };

  const ondeleteStudent = (rollNumber) => {
    if (window.confirm("Are you sure, you want to Delete this record?")) {
      deleteStudent(rollNumber).then(() => {
        setRefresh(!refresh);
      });
    }
  };

  const onEditBtnClick = (student) => {
    editStudent(student).then(() => {
      setRefresh(!refresh);
      setshowEditStudentForm(false);
    });
  };
  const addNewStudent = () => {
    setshowAddStudentForm(true);
  };
  const onEditStudent = (student) => {
    setRefresh(!refresh);
    setStudentToEdit(student);
    setshowEditStudentForm(true);
  };
  const onBackclick = () => {
    setshowAddStudentForm(false);
  };

  const onBtnclick = (student) => {
    const expression = /^[a-zA-Z0-9._-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,4}$/
    if(!expression.test(student.email)){
      window.confirm("Invalid Email Address");
      return;
    }
      addStudents(student).then(() => {
        setRefresh(!refresh);
        setshowAddStudentForm(false);
      });
    

    // addStudents(student).then(() => {
    //   setRefresh(!refresh);
    //   setshowAddStudentForm(false);
    // });
  };

  const renderApp = () => {
    if (error) {
      return <h2>Failed to load Students Data</h2>;
    } else if (showEditStudentForm) {
      return (
        <EditStudent
          student={studentToEdit}
          onEditBackclick={() => {
            setshowEditStudentForm(false);
          }}
          onEditStudent={onEditStudent}
          onEditBtnClick={onEditBtnClick}
        />
      );
    }
    if (showAddStudentForm) {
      return <AddStudent onBackclick={onBackclick} onBtnclick={onBtnclick} />;
    }
    return (
      <Students
        students={students}
        addNewStudent={addNewStudent}
        ondeleteStudent={ondeleteStudent}
        onEditStudent={onEditStudent}
      />
    );
  };

  return <div className="App">{renderApp()}</div>;
}

export default App;
