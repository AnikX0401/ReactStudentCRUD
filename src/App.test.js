import React from "react";
import {
  fireEvent,
  getByRole,
  getByText,
  render,
  screen,
  waitFor,
} from "@testing-library/react";
import App from "./App";
import {
  addStudents,
  getStudents,
  deleteStudent,
  editStudent,
  getStudentByRollNumber,
} from "./components/service";
import { mockStudents } from "./mockStudents";

jest.mock("./components/service", () => {
  return {
    getStudents: jest.fn(),
    addStudents: jest.fn(),
    deleteStudent: jest.fn(),
    editStudent: jest.fn(),
    getStudentByRollNumber: jest.fn(),
  };
});

describe("Students app", () => {
  beforeEach(() => {
    getStudents.mockResolvedValue([
      {
        name: "rahul",
        rollNumber: 1,
        trainings: ["JAVA", "HTML"],
        email: "rahul21@gmail.com",
      },
      {
        name: "manish",
        rollNumber: 2,
        trainings: ["JAVA", "HTML"],
        email: "manish21@gmail.com",
      },
      {
        name: "rohit",
        rollNumber: 3,
        trainings: ["HTML"],
        email: "rohit77@gmail.com",
      },
    ]);
  });

  it("render students data on fetching", async () => {
    const { asFragment } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders error message when students data is not fetched", async () => {
    getStudents.mockResolvedValue({ error: "failed" });

    render(<App />);

    await waitFor(() => {
      expect(
        screen.getByText("Failed to load Students Data")
      ).toBeInTheDocument();
    });
  });

  it("renders HomePage", () => {
    render(<App />);
    expect(screen.getByText("Home")).toBeInTheDocument();
    expect(screen.getByText("Add")).toBeInTheDocument();
  });

  it("renders error message on duplicate email", async () => {
    addStudents.mockResolvedValue({
      error: "Email address already exist..!",
    });

    render(<App />);

    const addBtn = screen.getByRole("link", {
      name: /add/i,
    });

    fireEvent.click(addBtn);
    await waitFor(() => {
      screen.getByLabelText("name:");
    });
    fireEvent.change(
      screen.getByRole("textbox", {
        name: /name:/i,
      }),
      { target: { value: "name789" } }
    );
    fireEvent.change(
      screen.getByRole("textbox", {
        name: /rollNumber:/i,
      }),
      { target: { value: "4" } }
    );
    fireEvent.change(
      screen.getByRole("textbox", {
        name: /trainings:/i,
      }),
      { target: { value: "HTML, CSS" } }
    );
    fireEvent.change(
      screen.getByRole("textbox", {
        name: /email:/i,
      }),
      { target: { value: "email@gmail.com" } }
    );
    const saveBtn = screen.getByRole("button", {
      name: /add/i,
    });

    fireEvent.click(saveBtn);
    await waitFor(() => {
      screen.getByText("Hey duplicate email id");
    });
    expect(addStudents).toHaveBeenCalledWith({
      email: "email@gmail.com",
      name: "name789",
      rollNumber: 4,
      trainings: ["HTML", "CSS"],
    });
  });

  it("deletes a student on clicking delete button", async () => {
    window.history.pushState({}, "hello", "/");
    getStudents.mockResolvedValue([
      {
        name: "rahul",
        rollNumber: 1,
        trainings: ["JAVA", "HTML"],
        email: "rahul21@gmail.com",
      },
      {
        name: "manish",
        rollNumber: 2,
        trainings: ["JAVA", "HTML"],
        email: "manish21@gmail.com",
      },
      {
        name: "rohit",
        rollNumber: 3,
        trainings: ["HTML"],
        email: "rohit77@gmail.com",
      },
    ]);

    deleteStudent.mockResolvedValue();
    render(<App />);
    await waitFor(() => {
      expect(screen.getByText("rahul")).toBeInTheDocument();
    });
    window.confirm = jest.fn().mockImplementation(() => true);

    fireEvent.click(screen.getAllByRole("button", { name: /delete/i })[0]);

    expect(window.confirm).toHaveBeenCalledWith(
      "Are you sure, you want to Delete this record?"
    );
    expect(deleteStudent).toHaveBeenCalledWith(1);
    expect(getStudents).toHaveBeenCalled();
  });

  it("renders Edit Student form on edit click and edit student on save", async () => {
    window.history.pushState({}, "hello", "/");
    getStudentByRollNumber.mockResolvedValue({
      name: "rahul",
      rollNumber: 1,
      trainings: ["JAVA", "HTML"],
      email: "rahul21@gmail.com",
    });
    render(<App />);
    await waitFor(() => {
      expect(screen.getByText("manish21@gmail.com")).toBeInTheDocument();
      //expect(screen.getByText(1)).toBeInTheDocument();
    });
    fireEvent.click(screen.getAllByRole("button", { name: /edit/i })[0]);
    await waitFor(() => {
      screen.getByLabelText("name:");
    });

    const rollNumberInput = screen.getByLabelText("rollNumber:");
    expect(rollNumberInput).toBeDisabled();

    fireEvent.change(
      screen.getByRole("textbox", {
        name: /name:/i,
      }),
      { target: { value: "New name" } }
    );

    fireEvent.change(
      screen.getByRole("textbox", {
        name: /trainings:/i,
      }),
      { target: { value: "JAVA, HTML, CSS" } }
    );

    editStudent.mockResolvedValue();

    fireEvent.click(screen.getByText("Save"));
    const save = screen.getByRole("button", {
      name: /save/i,
    });
    fireEvent.click(save);

    expect(editStudent).toHaveBeenCalledWith({
      name: "New name",
      rollNumber: 1,
      trainings: ["JAVA", "HTML", "CSS"],
    });
    expect(getStudentByRollNumber).toHaveBeenCalled();
  });

  it("filter students on search input", async () => {
    window.history.pushState({}, "hello", "/");
    getStudents.mockResolvedValue([
      {
        name: "rahul",
        rollNumber: 1,
        trainings: ["JAVA", "HTML"],
        email: "rahul21@gmail.com",
      },
      {
        name: "manish",
        rollNumber: 2,
        trainings: ["JAVA", "HTML"],
        email: "manish21@gmail.com",
      },
      {
        name: "rohit",
        rollNumber: 3,
        trainings: ["HTML"],
        email: "rohit77@gmail.com",
      },
    ]);
    render(<App />);
    fireEvent.change(screen.getByPlaceholderText("enter text here.."), {
      target: { value: "rahul" },
    });
    await waitFor(() => {
      expect(screen.getByText("rahul")).toBeInTheDocument();
      expect(screen.queryByText("manish")).not.toBeInTheDocument();
      expect(screen.queryByText("rohit")).not.toBeInTheDocument();
    });
    fireEvent.change(screen.getByPlaceholderText("enter text here.."), {
      target: { value: "" },
    });
    await waitFor(() => {
      expect(screen.getByText("rahul")).toBeInTheDocument();
      expect(screen.getByText("manish")).toBeInTheDocument();
      expect(screen.getByText("rohit")).toBeInTheDocument();
    });
    fireEvent.change(screen.getByPlaceholderText("enter text here.."), {
      target: { value: "2" },
    });
    await waitFor(() => {
      expect(screen.queryByText("rahul")).not.toBeInTheDocument();
      expect(screen.getByText("manish")).toBeInTheDocument();
      expect(screen.queryByText("rohit")).not.toBeInTheDocument();
    });
    fireEvent.change(screen.getByPlaceholderText("enter text here.."), {
      target: { value: "" },
    });
    await waitFor(() => {
      expect(screen.getByText("rahul")).toBeInTheDocument();
      expect(screen.getByText("manish")).toBeInTheDocument();
      expect(screen.getByText("rohit")).toBeInTheDocument();
    });
  });

  it("renders pagination buttons and changes page numbers", async () => {
    render(<App />);
    await waitFor(() => {
      expect(screen.getByText("rahul")).toBeInTheDocument();
    });

    fireEvent.click(
      screen.getByRole("button", {
        name: /next/i,
      })
    );

    fireEvent.click(
      screen.getByRole("button", {
        name: /previous/i,
      })
    );
    fireEvent.click(
      screen.getByRole("button", {
        name: /1/i,
      })
    );
  });

  it("renders previous button disabled on first page", async () => {
    render(<App />);
    await waitFor(() => {
      expect(screen.getByText("rahul")).toBeInTheDocument();
    });
    expect(
      screen.getByRole("button", {
        name: /previous/i,
      })
    ).toBeDisabled();
  });

  it("renders total students records based on set pagesize and next click", async () => {
    getStudents.mockResolvedValue(mockStudents);

    render(<App />);
    await waitFor(() => {
      expect(screen.getByText("rahul")).toBeInTheDocument();
    });

    for (let i = 1; i <= 10; i++) {
      expect(screen.getByText(mockStudents[i - 1].name)).toBeInTheDocument();
    }

    fireEvent.click(
      screen.getByRole("button", {
        name: /next/i,
      })
    );

    for (let i = 11; i <= 20; i++) {
      expect(screen.getByText(mockStudents[i - 1].name)).toBeInTheDocument();
    }
  });

  it("should highlight when trainings found of name JAVA", async () => {
    render(<App />);
    await waitFor(() => {
      expect(screen.getByText("rahul")).toBeInTheDocument();
    });
    const studentRow = await screen.findAllByRole("row");
    studentRow.forEach((student) => {
      if (student.textContent.includes("JAVA")) {
        expect(student).toHaveStyle("background-color: yellow");
      } else {
        expect(student).not.toHaveStyle("background-color: yellow");
      }
    });
  });
});
