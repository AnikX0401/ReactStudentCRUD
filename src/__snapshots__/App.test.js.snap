// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Students app render students data on fetching 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <nav>
      <ul>
        <li>
          <a
            href="/"
            style="color: rgb(12, 93, 199); padding: 5px; font-size: 25px;"
          >
            Home
          </a>
        </li>
        <li>
          <a
            href="/add"
            style="color: rgb(12, 93, 199); padding: 5px; font-size: 25px;"
          >
            Add
          </a>
        </li>
      </ul>
    </nav>
    <div
      class="student-list"
    >
      <h1>
        Students
      </h1>
      <input
        placeholder="enter text here.."
        type="text"
        value=""
      />
      <button>
        Search
      </button>
      <p>
        Total Students: 0
      </p>
      <table>
        <thead>
          <tr>
            <th>
              Name
            </th>
            <th>
              Roll Number
            </th>
            <th>
              Trainings
            </th>
            <th>
              Email
            </th>
            <th>
              Actions
            </th>
          </tr>
        </thead>
        <tbody />
      </table>
      <div
        class="pagination-buttons"
      >
        <button
          disabled=""
        >
          Previous
        </button>
        <button>
          Next
        </button>
      </div>
      <select>
        <option
          value="10"
        >
          10
        </option>
        <option
          value="20"
        >
          20
        </option>
        <option
          value="30"
        >
          30
        </option>
      </select>
      Paged: 0
    </div>
  </div>
</DocumentFragment>
`;
